prov/tcp/src/xnet_eq.c

```diff
@@ -158,57 +128,64 @@ static int xnet_eq_wait_try_func(void *arg)
 	return FI_SUCCESS;
 }
 
-int xnet_eq_add_progress(struct xnet_eq *eq, struct xnet_progress *progress,
-			 void *context)
+/* Safe to call even if domain was never added */
+static void xnet_eq_del_domain(struct xnet_eq *eq, struct xnet_domain *domain)
 {
-	return ofi_wait_add_fd(eq->util_eq.wait,
-			       ofi_dynpoll_get_fd(&progress->epoll_fd),
-			       POLLIN, xnet_eq_wait_try_func, NULL, context);
-}
+	assert(!xnet_progress_locked(&domain->progress));
 
-int xnet_eq_del_progress(struct xnet_eq *eq, struct xnet_progress *progress)
-{
-	return ofi_wait_del_fd(eq->util_eq.wait,
-			       ofi_dynpoll_get_fd(&progress->epoll_fd));
+	ofi_mutex_lock(&eq->domain_lock);
+	fid_list_remove(&eq->domain_list, NULL,
+			&domain->util_domain.domain_fid.fid);
+
+	if (eq->util_eq.wait) {
+		(void) ofi_wait_del_fd(eq->util_eq.wait,
+				ofi_dynpoll_get_fd(&domain->progress.epoll_fd));
+	}
+	ofi_mutex_unlock(&eq->domain_lock);
 }
 
-static int xnet_eq_monitor_all(struct xnet_eq *eq, struct xnet_fabric *fabric)
+void xnet_del_domain_progress(struct xnet_domain *domain)
 {
-	struct xnet_domain *domain;
-	struct dlist_entry *error_item;
+	struct xnet_fabric *fabric;
+	struct xnet_eq *eq;
 	struct dlist_entry *item;
-	int ret;
 
-	ret = xnet_eq_add_progress(eq, &fabric->progress,
-				   &fabric->util_fabric.fabric_fid);
-	if (ret)
-	    return ret;
+	fabric = container_of(domain->util_domain.fabric, struct xnet_fabric,
+			      util_fabric.fabric_fid);
 
 	ofi_mutex_lock(&fabric->util_fabric.lock);
-	dlist_foreach(&fabric->util_fabric.domain_list, item) {
-		domain = container_of(item, struct xnet_domain,
-				      util_domain.list_entry);
-		ret = xnet_eq_add_progress(eq, &domain->progress,
-					   &domain->util_domain.domain_fid);
-		if (ret) {
-			error_item = item;
-			goto clean;
-		}
+	dlist_foreach(&fabric->eq_list, item) {
+		eq = container_of(item, struct xnet_eq, fabric_entry);
+		xnet_eq_del_domain(eq, domain);
 	}
-	dlist_insert_tail(&eq->wait_eq_entry, &fabric->wait_eq_list);
 	ofi_mutex_unlock(&fabric->util_fabric.lock);
-	return FI_SUCCESS;
+}
+
+int xnet_add_domain_progress(struct xnet_eq *eq, struct xnet_domain *domain)
+{
+	int ret;
 
-clean:
-	/* Traverse the list backwards from where the error occurred */
-	dlist_foreach_reverse(error_item, item) {
-		domain = container_of(item, struct xnet_domain,
-				      util_domain.list_entry);
-		xnet_eq_del_progress(eq, &domain->progress);
+	assert(!xnet_progress_locked(&domain->progress));
```

Comment:
I tested your PR and it immediately triggers this assertion failure. This is expected.
Here, you want to verify that the current thread doesn't hold the progress lock in order to avoid a deadlock later in `xnet_start_progress()`. However, `xnet_progress_locked()` checks if the lock is acquired by anyone, not only the current thread. The lock might be acquired by another thread when the assertion failure is executed. 

After I remove this line (and the one line 134), everything works fine for my use case.

Reply:
Excellent point, and yes, you understood why I added this check.  I'll update with the check removed.  I don't know if there's a way to check against the current thread only.